torch/lib/TensorLib/templates/TensorDerived.cpp

```diff
@@ -0,0 +1,67 @@
+#include "TensorLib/${Tensor}.h"
+#include "TensorLib/HalfConvert.h"
+
+//sketch:
+
+/*
+
+class Type {
+  // ctor is configurable as well
+  Tensor new(...) {}
+  // all  methods and pure functions, virtually dispatched
+  virtual Tensor add(Tensor a, Tensor b);
+
+  // no Tensor arguments but virtually dispatched on type
+  virtual Tensor linspace(int a, int b);
+  //
+}
+
+// base class of derived tensors
+class TensorImpl {
+  Type * type_; //non-virtual so that we don't double indirect unnecessarily ...
+
+  // not virtual - Type * handles the virtual dispatch
+  Tensor add(Tensor b) {
+    type_->add(self,b);
+  }
+}
```

Comment:
Right now, for simplicity -- all things that actually unwrap the Tensors are defined in Type. It is less logic in the wrapper generator, and fewer places where we have to generate declarations.